### PR TITLE
schema.Reload(): ignore column reading errors for views only, error for tables

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Followup and refine to https://github.com/vitessio/vitess/pull/13421.

In https://github.com/vitessio/vitess/pull/13421 we ignore (log and continue) errors while reading columns for tables/views. This is too permissive and might cause undesired behavior down the road for code that uses the caches table data. An error reading table columns should remain an error and fail the `Reload()` operation. For views, we can indeed ignore such failure, for two reasons:

1. It is possible to have an invalid view definition in MySQL. So while highly undesired, this is a "valid" situation that we should accommodate.
2. None of our internal processes cares much about view columns (e.g. vreplication doesn't read from views).

So this PR only ignores the errors for views, and does error when failing to read table columns.

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/13421
- https://github.com/vitessio/vitess/issues/13422

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/schema/engine_test.go [hunk 1]
```diff
@@ -447,18 +447,16 @@ func TestOpenFailedDueToLoadTableErr(t *testing.T) {
 	db.AddQueryPattern(fmt.Sprintf(mysql.GetColumnNamesQueryPatternForTable, "test_view"),
 		sqltypes.MakeTestResult(sqltypes.MakeTestFields("column_name", "varchar"), ""))
 	// rejecting the impossible query
-	db.AddRejectedQuery("SELECT * FROM `fakesqldb`.`test_view` WHERE 1 != 1", errors.New("The user specified as a definer ('root'@'%') does not exist (errno 1449) (sqlstate HY000)"))
+	db.AddRejectedQuery("SELECT * FROM `fakesqldb`.`test_view` WHERE 1 != 1", mysql.NewSQLErrorFromError(errors.New("The user specified as a definer ('root'@'%') does not exist (errno 1449) (sqlstate HY000)")))
 
 	AddFakeInnoDBReadRowsResult(db, 0)
 	se := newEngine(10, 1*time.Second, 1*time.Second, 0, db)
 	err := se.Open()
 	// failed load should not return any error, instead should be logged.
```
Comment (GuptaManan100) [inline go/vt/vttablet/tabletserver/schema/engine_test.go:455]: This comment has to change

**Review by GuptaManan100 — APPROVED** [review]
LGTM

File: go/vt/vttablet/tabletserver/schema/engine_test.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/schema/engine_test.go:455]: Updated comment

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/vt/vttablet/tabletserver/schema/engine.go [hunk 2]
```diff
@@ -491,9 +497,24 @@ func (se *Engine) reload(ctx context.Context, includeStats bool) error {
...
+			var emptyColumnsError mysqlctl.EmptyColumnsErr
+			if errors.As(err, &emptyColumnsError) && isView {
+				log.Warningf("Failed reading schema for the table: %s, error: %v", tableName, err)
+				continue
+			}
+			sqlErr, isSQLErr := mysql.NewSQLErrorFromError(err).(*mysql.SQLError)
+			if isSQLErr && sqlErr != nil && sqlErr.Number() == mysql.ERNoSuchUser && isView {
+				// A VIEW that has an invalid DEFINER, leading to:
+				// ERROR 1449 (HY000): The user specified as a definer (...) does not exist
+				log.Warningf("Failed reading schema for the table: %s, error: %v", tableName, err)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/schema/engine.go:513]: The scenario you're refrring to is covered by the fact the query on `information_schema` does not return a row. It is handled here: https://github.com/vitessio/vitess/pull/13442/files#diff-106487fd3d9cbbb2b783c2d8aff23d3f448de495303a13801c6eebdbcefc0c7aR311

**Comment by shlomi-noach** [comment]
Similarly to https://github.com/vitessio/vitess/pull/13421, this will be backported to `15`, `16`, `17`.

**Comment by shlomi-noach** [comment]
Sadly, the backport will be manual.

File: go/vt/vttablet/tabletserver/schema/engine.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/vttablet/tabletserver/schema/engine.go:513]: Following up on this, do we have a unit test for this case?

**Pull request merged**